Make speech to text fully asynchronous

rute-santos opened the pull request:
Whisper can now run fully asynchronously i.e. a workflow starts the subtitle generation and finishes execution; the generated subtitles are attached later by running another workflow.

This is useful when we want to publish the recording quickly without having to wait for the subtitles, and attach them later.

Changes:

- New service that runs on the admin node and tracks those jobs (new db table).
- Make sure input files and arguments are not deleted at the end of the starting workflow.
- Add quartz job to schedule workflow to attach subtitles when all STT jobs have finished.

Closes #6967.


### How to test this patch

1 - Create a workflow to attach the subtitles. Operation example:

```
- id: speechtotext-attach
  description: Attach subtitles
  configurations:
    - target-flavor: captions/delivery
    - target-element: track
    - target-tags: engage-download

- id: publish-engage-aws
  description: Publish media package
  configurations:
    - download-source-flavors: captions/delivery,security/*
    - strategy: merge
```

2 - Add it to the quartz job configuration in etc/org.opencastproject.speechtotext.async.impl.SpeechToTextWorkflowSchedulerQuartz

3 - Run a workflow that starts Whisper transcriptios after the media package is published. Operation example:

```
- id: speechtotext
  description: Starts Whisper transcription
  configurations:
    - source-flavor: '*/delivery'
    - target-flavor: captions/delivery
    - limit-to-one: true
    - track-selection-strategy: try_presenter_first
    - async: true
```

4 - Verify that the subtitles are attached to the media package after the second workflow ran.

rute-santos committed:
Make speech to text fully asynchronous.

Add table to track STT async jobs
Add service to track/untrack async jobs
Make sure input files and arguments are not deleted at the end of the workflow by using parentless jobs.
Add quartz job to schedule workflow to attach subtitles when all STT jobs have finished
Update speech-to-text and speech-to-text-attach woh to invoke service to track/untrack jobs
Update documentation.

lkiesow commented on modules/speech-to-text-async-impl/pom.xml:
```diff
@@ -0,0 +1,161 @@
+      <groupId>org.eclipse.persistence</groupId>
+      <artifactId>jakarta.persistence</artifactId>
+    </dependency>
+    <!-- dependency>
```
This seems like an accidental left over?

rute-santos replied:
Yes, removing...

rute-santos committed:
Remove leftover comments.

gregorydlogan commented on modules/speech-to-text-async-impl/src/main/java/org/opencastproject/speechtotext/async/impl/SpeechToTextWorkflowSchedulerQuartz.java:
```diff
@@ -0,0 +1,403 @@
+ */
+@Component(property = { "service.description=Speech to Text Workflow Scheduler" }, immediate = true, service = {
+        ManagedService.class })
+public class SpeechToTextWorkflowSchedulerQuartz extends AbstractScanner implements ManagedService {
```
```suggestion
public class SpeechToTextWorkflowSchedulerQuartz extends AbstractScanner {
```

gregorydlogan replied:
`ManagedService` should be avoided if possible in favour of declarative services

rute-santos replied:
I tried and I wasn't successful with that. All other classes that extend AbstractScanner also implement ManagedService so maybe this is a change that needs to be done to all at once?

gregorydlogan replied:
Ok, fair. Didn't catch that. It's in my list to remove ManagedService, so let's ignore this and I'll deal with it later.

gregorydlogan commented on modules/speech-to-text-async-impl/src/main/java/org/opencastproject/speechtotext/async/impl/SpeechToTextWorkflowSchedulerQuartz.java:
```diff
@@ -0,0 +1,403 @@
+    }
+  }
+
+  @Override
+  public void updated(Dictionary<String, ?> properties) throws ConfigurationException {
```
Because we're not doing `ManagedService`
```suggestion
  @Modified
  public void modified(Map<String, Object> properties) {
```

gregorydlogan commented on modules/speech-to-text-async-impl/src/main/java/org/opencastproject/speechtotext/async/impl/SpeechToTextWorkflowSchedulerQuartz.java:
```diff
@@ -0,0 +1,403 @@
+
+import org.apache.commons.lang3.StringUtils;
+import org.osgi.service.cm.ConfigurationException;
+import org.osgi.service.cm.ManagedService;
```
```suggestion
```

gregorydlogan commented on modules/speech-to-text-async-impl/src/main/java/org/opencastproject/speechtotext/async/impl/SpeechToTextAsyncTrackerImpl.java:
```diff
@@ -0,0 +1,67 @@
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+@Component(property = { "service.description=Speech to Text Async Tracker" }, immediate = true, service = {
```
Can we reformat this so it reads nicer?

gregorydlogan commented on assemblies/karaf-features/src/main/feature/feature.xml:
You've added both an API and an impl bundle here, which isn't *wrong*, but are we likely to see a different implementation of this API at any point?  I would be tempted to collapse them together.

rute-santos replied:
I always separate them... but please let me know if you feel strongly about this and I will change :)

gregorydlogan commented on modules/speech-to-text-async-impl/src/main/java/org/opencastproject/speechtotext/async/impl/SpeechToTextWorkflowSchedulerQuartz.java:
```diff
@@ -0,0 +1,403 @@
+      setCronExpression(cronExpression);
+      logger.info("{} : {}", PARAM_KEY_CRON_EXPR, cronExpression);
+
+      attachWorkflowDef = (String) properties.get(WORKFLOW);
```
So if this key is unset we just roll with it?  What's the usecase for *not* having a workflow ID set?

rute-santos replied:
Fixed.

gregorydlogan commented on modules/speech-to-text-async-impl/src/main/java/org/opencastproject/speechtotext/async/impl/SpeechToTextWorkflowSchedulerQuartz.java:
```diff
@@ -0,0 +1,403 @@
+   * @throws SpeechToTextAsyncException
+   */
+  void handleTranscriptionFinished() throws SpeechToTextAsyncException {
+    if (attachWorkflowDef == null) {
```
If we enforce this being set during `activate` we don't need to check it here.

gregorydlogan commented on modules/speech-to-text-async-impl/src/main/java/org/opencastproject/speechtotext/async/impl/SpeechToTextWorkflowSchedulerQuartz.java:
```diff
@@ -0,0 +1,403 @@
+              .applyWorkflowToLatestVersion(mpIds, ConfiguredWorkflow.workflow(wfDef, wfConfig));
+      WorkflowInstance wf = wfList.size() > 0 ? wfList.get(0) : null;
+      logger.info("Workflow {} {}started for mp {}{}", attachWorkflowDef, wf != null ? "" : "NOT ", mpId,
+              wf != null ? ": " + wf.getId() : "");
```
While this is more verbose, it makes for *way* easier reading
```suggestion
      if (wf != null) {
        logger.info("Workflow {} started for mp {}: {}", attachWorkflowDef, mpId, wf.getId());
      } else {
        logger.info("Workflow {} NOT started for mp {}", attachWorkflowDef, mpId);
      }
```

rute-santos replied:
Fixed.

gregorydlogan reviewed (changes requested):
A few issues, the biggest one being that the unit tests are failing for me:

```
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.169 s <<< FAILURE! -- in org.opencastproject.speechtotext.impl.Speec
hToTextServiceImplTest                                                                                                                        
[ERROR] org.opencastproject.speechtotext.impl.SpeechToTextServiceImplTest.testVoskSpeechToText -- Time elapsed: 0.080 s <<< FAILURE!          
java.lang.AssertionError:                                                                                                                     
                                                                                                                                              
  Unexpected method call EasyMock for interface org.opencastproject.serviceregistry.api.ServiceRegistry -> ServiceRegistry.getCurrentJob()    
        at org.easymock.internal.MockInvocationHandler.invoke(MockInvocationHandler.java:46)                                                  
        at org.easymock.internal.ObjectMethodsFilter.invoke(ObjectMethodsFilter.java:100)                                                     
        at jdk.proxy2/jdk.proxy2.$Proxy15.getCurrentJob(Unknown Source)                                                                       
        at org.opencastproject.speechtotext.impl.SpeechToTextServiceImpl.transcribe(SpeechToTextServiceImpl.java:189)                         
        at org.opencastproject.speechtotext.impl.SpeechToTextServiceImplTest.testVoskSpeechToText(SpeechToTextServiceImplTest.java:152)
[...]
```

rute-santos committed:
Address reviewer comments.

Fix vosk unit test.
Fix formatting.
Throw exception if attach workflow not set.
Rebased on current develop

rute-santos committed:
Improve Whisper async error handling

Detect Whisper failure before starting workflow to attach subtitles.
Add an optional retry workflow.
Also retry jobs stuck in progress for a long time.
Do not fail the operation if workspace cleanup fails

gregorydlogan commented on modules/speech-to-text-async-impl/src/main/java/org/opencastproject/speechtotext/async/impl/SpeechToTextWorkflowSchedulerQuartz.java:
```diff
@@ -0,0 +1,494 @@
+      setEnabled(properties.get(PARAM_KEY_ENABLED) != null
+          && Boolean.valueOf(properties.get(PARAM_KEY_ENABLED).toString()));
+      if (!isEnabled()) {
+        logger.info("Speech to text workflow scheduler quartz job is disabled");
```
Why would you ever want this to be disabled?  Allowing the user to modify this seems like it can only lead to pain.

rute-santos replied:
When Whisper is run synchronously, which I believe is the majority of cases, you don't need a quartz job that does nothing running all the time.

gregorydlogan commented on modules/speech-to-text-async-impl/src/main/java/org/opencastproject/speechtotext/async/impl/SpeechToTextWorkflowSchedulerQuartz.java:
```diff
@@ -0,0 +1,494 @@
+          abandonAfterMs = Integer.parseInt((String) properties.get(ABANDON_AFTER_SECS)) * 1000;
+        } catch (Exception e) {
+          // Log and use default
+          logger.warn("Invalid configuration for {}: '{}'", ABANDON_AFTER_SECS, properties.get(ABANDON_AFTER_SECS), e);
```
Should list the fallback default as well.

```suggestion
          logger.warn("Invalid configuration for {}: '{}', defaulting to '{}'", ABANDON_AFTER_SECS, properties.get(ABANDON_AFTER_SECS), DEFAULT_ABANDON_AFTER_SECSe);
```

rute-santos replied:
Done!

gregorydlogan commented on modules/speech-to-text-async-impl/src/main/java/org/opencastproject/speechtotext/async/impl/SpeechToTextWorkflowSchedulerQuartz.java:
```diff
@@ -0,0 +1,494 @@
+
+    try {
+      // STTs that are stuck in progress for a long time will be transitioned to error so that they will be retried.
+      database.transitionStatusByDate(SpeechToTextControl.Status.TranscriptionError, nowMinusInterval,
+              SpeechToTextControl.Status.InProgress);
+
+      // STTs that had a workflow to attach subtitles started are transitioned to done because afaik everything worked
+      // as supposed to
+      database.transitionStatusByDate(SpeechToTextControl.Status.Done, nowMinusInterval,
+              SpeechToTextControl.Status.WorkflowInProgress);
```
The count here should be logged.  Ideally we'd log the individual control/mediapackage/workflow ID at debug too, but looking at the queries I'm not sure how reasonable that is.

rute-santos replied:
Added debug logging with counter.

gregorydlogan commented on modules/speech-to-text-impl/src/main/java/org/opencastproject/speechtotext/impl/endpoint/SpeechToTextServiceRestEndpoint.java:
```diff
@@ -102,7 +102,9 @@ public class SpeechToTextServiceRestEndpoint extends AbstractJobProducerEndpoint
-              description = "Enable translation to english") },
+              description = "Enable translation to english"),
+          @RestParameter(name = "async", isRequired = false, type = BOOLEAN,
+              description = "If job will run outside the scope of a workflow") },
```
Include the default (false?) too please.

rute-santos replied:
Done!

gregorydlogan commented on modules/speech-to-text-workflowoperation/src/main/java/org/opencastproject/workflow/handler/speechtotext/SpeechToTextWorkflowOperationHandler.java:
```diff
@@ -388,11 +394,31 @@ private void createSubtitleAsync(WorkflowInstance workflow, List<Track> tracks,
+        String fileName = String.format("%d-%s%s", workflow.getId(), track.getIdentifier(), extension);
+        URI inputUri = workspace.putInCollection(STT_ASYNC_COLLECTION, fileName, workspace.read(track.getURI()));
+
+        var job = speechToTextService.transcribe(inputUri, languageCode, translate, true);
```
We default to doing this async, but we don't have any changes to the workflows to doing the actual attach?  Or am I misunderstanding something?

rute-santos replied:
There are 2 ways to run it asynchronously:

1. Have the speechtotext woh and speechtotext-attach woh in the same workflow - this is how it was implemented before and was kept as is. The speechtotext-attach woh in the same workflow will wait until subtitles are ready.
2. Have the speechtotext woh in one workflow and the speechtotext-attach in another, started by the quartz scheduler action when the subtitles are ready - this is what this PR implements.

I didn't provide a sample workflow for that; I only updated the documentation (whisper.md) as it seems there's no sample workflow that uses speechtotext? Please let me know if this is not enough!

gregorydlogan reviewed (changes requested):
This needs a release notes entry so it gets picked up on by the RMs as well.

rute-santos committed:
Address reviewer comments. Improve logging. Add to release notes.

rute-santos committed:
Merge remote-tracking branch 'opencast-community/develop' into whisper-async-6967
